src/apple/disk.rs

```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
 
     fn refresh(&mut self) -> bool {
         unsafe {
-            let mut stat: statfs = mem::zeroed();
-            let mount_point_cpath = to_cpath(&self.mount_point);
-            if statfs(mount_point_cpath.as_ptr() as *const i8, &mut stat) == 0 {
-                self.available_space = u64::from(stat.f_bsize).saturating_mul(stat.f_bavail);
-                true
+            let requested_properties = build_requested_properties(&[
+                ffi::kCFURLVolumeAvailableCapacityKey,
+                ffi::kCFURLVolumeAvailableCapacityForImportantUsageKey,
+            ]);
+            match get_disk_properties(&self.volume_url, &requested_properties) {
+                Some(disk_props) => {
+                    self.available_space = get_available_volume_space(&disk_props);
+                    true
+                }
+                None => false,
+            }
+        }
+    }
+}
+
+pub(super) fn get_disks() -> Vec<Disk> {
+    let raw_disks = unsafe {
+        let count = libc::getfsstat(ptr::null_mut(), 0, libc::MNT_NOWAIT);
+        if count < 1 {
+            return Vec::new();
+        }
+        let bufsize = count * std::mem::size_of::<libc::statfs>() as libc::c_int;
+        let mut disks = Vec::with_capacity(count as _);
+        let count = libc::getfsstat(disks.as_mut_ptr(), bufsize, libc::MNT_NOWAIT);
+
+        if count < 1 {
+            return Vec::new();
+        }
+
+        disks.set_len(count as usize);
+
+        disks
+    };
+
+    // Create a list of properties about the disk that we want to fetch.
+    let requested_properties = unsafe {
+        build_requested_properties(&[
+            ffi::kCFURLVolumeIsEjectableKey,
+            ffi::kCFURLVolumeIsRemovableKey,
+            ffi::kCFURLVolumeIsInternalKey,
+            ffi::kCFURLVolumeTotalCapacityKey,
+            ffi::kCFURLVolumeAvailableCapacityForImportantUsageKey,
+            ffi::kCFURLVolumeAvailableCapacityKey,
+            ffi::kCFURLVolumeNameKey,
+            ffi::kCFURLVolumeIsBrowsableKey,
+            ffi::kCFURLVolumeIsLocalKey,
+        ])
+    };
+
+    let mut disks = Vec::with_capacity(raw_disks.len());
+    for c_disk in raw_disks {
+        let volume_url = CFReleaser::new(unsafe {
+            core_foundation_sys::url::CFURLCreateFromFileSystemRepresentation(
+                kCFAllocatorDefault,
+                c_disk.f_mntonname.as_ptr() as *const u8,
+                c_disk.f_mntonname.len() as isize,
+                false as u8,
+            )
+        })
+        .unwrap();
```

Comment:
Please use a `match` and add a `sysinfo_debug!` call in case it failed. I'm really uneasy to just `unwrap`.